drivers/rtc/rtc_sam.c

```diff
@@ -0,0 +1,681 @@
+
+	if (data->alarm_callback) {
+		config->regs->RTC_IER = RTC_IER_ALREN;
+	} else {
+		config->regs->RTC_IDR = RTC_IDR_ALRDIS;
+	}
+
+	irq_enable(config->irq_num);
+	k_spin_unlock(&data->lock, key);
+	return 0;
+}
+#endif /* CONFIG_RTC_ALARM */
+
+#ifdef CONFIG_RTC_UPDATE
+static int rtc_sam_update_set_callback(const struct device *dev, rtc_update_callback callback,
+				       void *user_data)
+{
+	struct rtc_sam_data *data = dev->data;
+
+	const struct rtc_sam_config *config = dev->config;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	irq_disable(config->irq_num);
+
+	data->update_callback = callback;
+	data->update_user_data = user_data;
+
+	if (data->update_callback) {
+		config->regs->RTC_IER = RTC_IER_SECEN;
+	} else {
+		config->regs->RTC_IDR = RTC_IDR_SECDIS;
+	}
+
+	irq_enable(config->irq_num);
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+#endif /* CONFIG_RTC_UPDATE */
+
+#ifdef CONFIG_RTC_CALIBRATION
+static int rtc_sam_set_calibration(const struct device *dev, int32_t calibration)
+{
+	struct rtc_sam_data *data = dev->data;
+	const struct rtc_sam_config *config = dev->config;
+	bool negative_calibration;
+	bool high_calibration;
+	uint32_t slow_clock_calibration;
+	uint32_t mr;
+
+	if ((calibration < RTC_SAM_CALIBRATE_PPB_MIN) ||
+	    (calibration > RTC_SAM_CALIBRATE_PPB_MAX)) {
+		return -EINVAL;
+	}
+
+	/* The value written to the register is absolute */
+	if (calibration < 0) {
+		negative_calibration = true;
+		calibration = -calibration;
+	} else {
+		negative_calibration = false;
+	}
+
+	if (calibration < RTC_SAM_CALIBRATE_PPB_QUANTA) {
+		high_calibration = false;
+		slow_clock_calibration = 0;
+	} else if (calibration < RTC_SAM_CALIBRATE_PPB_LOW_SCALE) {
+		high_calibration = false;
+		slow_clock_calibration = (uint32_t)((3906000 / (20 * calibration)) - 1);
+	} else {
+		high_calibration = true;
+		slow_clock_calibration = (uint32_t)((3906000 / calibration) - 1);
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	rtc_sam_disable_wp();
+
+	mr = config->regs->RTC_MR;
+
+	if (negative_calibration == true) {
+		mr |= RTC_MR_NEGPPM;
+	} else {
+		mr &= ~RTC_MR_NEGPPM;
+	}
+
+	mr &= ~RTC_MR_CORRECTION_Msk;
+	mr |= RTC_MR_CORRECTION(slow_clock_calibration);
+
+	if (high_calibration == true) {
+		mr |= RTC_MR_HIGHPPM;
+	} else {
+		mr &= ~RTC_MR_HIGHPPM;
+	}
+
+	config->regs->RTC_MR = mr;
+
+	rtc_sam_enable_wp();
+
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int rtc_sam_get_calibration(const struct device *dev, int32_t *calibration)
+{
+	const struct rtc_sam_config *config = dev->config;
+
+	uint32_t mr;
+	int32_t correction;
+
+	if (calibration == NULL) {
+		return -EINVAL;
+	}
+
+	mr = config->regs->RTC_MR;
+
+	correction = (int32_t)RTC_SAM_REG_GET_FIELD(mr, MR_CORRECTION);
+
+	if (correction == 0) {
+		*calibration = 0;
+	} else if (mr & RTC_MR_HIGHPPM) {
+		*calibration = 3906000 / (correction + 1);
+	} else {
+		*calibration = 3906000 / ((correction + 1) * 20);
+	}
```

Comment:
ditto

Reply:
covered by previous point, added a link to the formula
```
/* Formula documented in rtc_sam_set_calibration() */
```